### PR TITLE
add test_iot_sdio.c

### PR DESCRIPTION
**Pull request by qizengqizeng**
add test_iot_sdio.c
Description
-----------

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 1]
```diff
@@ -0,0 +1,714 @@
+/*
+ * Amazon FreeRTOS Common IO V1.0.0
+ * Copyright (C) 2019 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/include/iot_sdio.h]: 2020

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 2]
```diff
@@ -0,0 +1,1042 @@
...
+/**
+ * @brief user callback function
+ */
+void prvSdioCallback( IotSdioSlotHandle_t const pxSdioHandle,
+                      IotSdioEventType_t eSdioEvent, void * pvUserContext )
+{
+    switch (eSdioEvent)
+    {
+        case eSdioCardInterruptEvent:
+            printf("sdiocallback with eSdioCardInterruptEvent\n");
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: No printfs in tests.

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 3]
```diff
@@ -0,0 +1,1042 @@
...
+ *
+ */
+TEST( TEST_IOT_SDIO, AFQP_IotSdioOpenClose )
+{
+    IotSdioSlotHandle_t xSdioHandle = NULL;
+
+    int32_t lRetVal;
+
+    /* Open with valid host index and slot index should return a non NULL handle. */
+    xSdioHandle = iot_sdio_open(uctestIotSdioValidHostIdx,
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: Can you use a global handle if present? If not open it? Please see I2C test as an example

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 4]
```diff
@@ -0,0 +1,1042 @@
...
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    lRetVal = iot_sdio_ioctl(xSdioHandle, eSDIOSetCardDetectParams, (void *)&xtestIotSdioCardDetectParam);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);;
+
+    xCardPowerCtl.bPowerOn = false;
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOPowerOnOffCard,
+                             (void *)&xCardPowerCtl);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: What does this IOCTL validate? Is there a way to check if the IOCTL did something other than just return SUCCESS?

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 5]
```diff
@@ -0,0 +1,1042 @@
...
+                             eSDIOCheckCardPresence,
+                             (void *)&bCardPresent);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+    TEST_ASSERT_EQUAL( true, bCardPresent);
+
+    xCardPowerCtl.bPowerOn = true;
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOPowerOnOffCard,
+                             (void *)&xCardPowerCtl);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: Same as #382. If the validation is for connect to be SUCCESS. May be just add a comment

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 6]
```diff
@@ -0,0 +1,1042 @@
...
+
+    lRetVal = iot_sdio_card_connect(xSdioHandle);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    /* Switch to high speed */
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOSwitchToHighSpeed,
+                             NULL);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: Is this a must to support?

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 7]
```diff
@@ -0,0 +1,1042 @@
...
+                                testIotSdio_EHS_ADDR,
+                                (uint8_t *)&ucReadVal);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+    TEST_ASSERT_EQUAL(testIotSdio_EHS_MASK, testIotSdio_EHS_MASK&ucReadVal);
+
+    /* Set driver strength */
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOSetDriverStrength,
+                             (void *)&etestIotSdioDriverType);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: Is this a must to support? Can the implementation return NOT_SUPPORTED?

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 8]
```diff
@@ -0,0 +1,1042 @@
...
+    TEST_ASSERT_EQUAL(0, testIotSdio_IO1_EN_MASK & ucReadVal);
+
+
+    /* Set card inactive */
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOSetCardInactive,
+                             NULL);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    if (btestIotSdioVerifyInactive) {
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: q: Why is this guarded with a variable?

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 4 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: Some pluggable card require card to be off state for proper detection. For embedded SDIO device, it simply return SUCCESS.

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 3 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: will update.

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 2 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: will remove.

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 1 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: Will update.

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 4 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: Will update comment

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 5 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: will update comment

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 6 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: Host must support. Card may not support. Will add comprehensive card side check.

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 7 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: Yes. Card side may selectively un-support some type. The following check is comprehensive to cover each case.

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 8 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: Some card respond to the command with SUCCESS response but never go into inactive state. This variable allows config the DUT to go through this check or skip.

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 1 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: done

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 2 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: done

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 3 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: done

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 4 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: done

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 5 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: done

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 6 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: done

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 9]
```diff
@@ -0,0 +1,714 @@
...
+/**
+ * @brief IOCTL request types.
+ */
+typedef enum IotSdioIoctlRequest
+{
+    eSDIOSetDriverStrength,   /*!< Set driver strength type
+                               *   input: SDIO card driver strength.
+                               *          input data type IotSdioSdDriverStrength_t */
+    eSDIOSetCardInactive,     /*!< put card in inactive state */
+    eSDIOSwitchToHighSpeed,   /*!< switch to high speed mode if card spports it */
```
Comment (lundinc2) [inline libraries/abstractions/common_io/include/iot_sdio.h]: nit: typo on supports

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 10]
```diff
@@ -0,0 +1,714 @@
...
+                               *   input: I/O function number and boolean to indicate enable
+                               *          or disable, input data type IotSdioPerFuncEnable_t */
+    eSDIOSelectIo,            /*!< Select an I/O function for suspend/resume
+                               *   input: I/O function number, input data type IotSdioFunction_t */
+    eSDIOAbortIo,             /*!< Abort an ongoing transfer initiated by the specified I/O function
+                               *   input: I/O function number. Input data type IotSdioFunction_t */
+    eSDIOReadCis,             /*!< Host read list of tuples from card CIS
+                               *   input: I/O function number and tuple list to read,
+                               *          input data type IotSdioReadCis_t*/
+    eSDIOPowerOnOffCard,      /*!< Power on/off card
```
Comment (lundinc2) [inline libraries/abstractions/common_io/include/iot_sdio.h]: Why does the boolean true indicate power on and use the boards default power method? Is there a missing enum?

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 11]
```diff
@@ -0,0 +1,714 @@
...
+    bool bUseCallback; /*!< indicate if user callback should be used to power on/off card. */
+} IoTSdioPowerOnOffCard_t;
+
+/**
+ * @brief input parameter type for eSDIOSetFuncBlockSize ioctl request
+ */
+typedef struct IotSdioFuncBlkSize
+{
+    IotSdioFunction_t eFunc; /*!< IO function number range from eSdioFunction0 to eSdioFunction7*/
+    uint32_t ulBlockSize;    /*!< Block size to be set for I/O function. Must not begreater
```
Comment (lundinc2) [inline libraries/abstractions/common_io/include/iot_sdio.h]: nit: begreater

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 12]
```diff
@@ -0,0 +1,714 @@
...
+ *                              to eSdioFunction7.
+ * @param[in]   ulRegAddr       The address of the byte of data inside of the
+ *                              selected function to read.
+ *                              Range is 0 - 0x1ffff.
+ * @param[out]   pucData        Contains data read when command succeeded.
+ *
+ * @return
+ *   - IOT_SDIO_SUCCESS on success.
+ *   - IOT_SDIO_INVALID_VALUE if any parameter is invalid.
+ *   - IOT_SDIO_TRANSFER_FAIL if failed to send command to for receive
```
Comment (lundinc2) [inline libraries/abstractions/common_io/include/iot_sdio.h]: The phrasing is a bit confusing here

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+ *                              When bBlockMode is false, a value of 0 shall
+ *                              cause 512 bytes to be written.
+ *                              When bBlockMode is true, a value of 0 shall
+ *                              cause infinite blocks to be written until
+ *                              iot_sdio_abort_io is called.
+ *
+ * @return
+ *   - IOT_SDIO_SUCCESS on success.
+ *   - IOT_SDIO_INVALID_VALUE if any parameter is invalid.
+ *   - IOT_SDIO_TRANSFER_FAIL if failed to send command to for receive
```
Comment (lundinc2) [inline libraries/abstractions/common_io/include/iot_sdio.h]: The phrasing is confusing here as well

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 14]
```diff
@@ -0,0 +1,714 @@
...
+ *                              When bBlockMode is false, a value of 0 shall
+ *                              cause 512 bytes to be read.
+ *                              When bBlockMode is true, a value of 0 shall
+ *                              cause infinite blocks to be read until
+ *                              iot_sdio_abort_io is called.
+ *
+ * @return
+ *   - IOT_SDIO_SUCCESS on success.
+ *   - IOT_SDIO_INVALID_VALUE if any parameter is invalid.
+ *   - IOT_SDIO_TRANSFER_FAIL if failed to send command to for receive
```
Comment (lundinc2) [inline libraries/abstractions/common_io/include/iot_sdio.h]: Same phrase here

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 15]
```diff
@@ -0,0 +1,1073 @@
...
+        gIotSdioHandle[ testIotSdio_HOST_INSTANCE ][ testIotSdio_SLOT_INSTANCE ] = NULL;
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Tear down function called after each test in this group is executed.
+ */
+TEST_TEAR_DOWN( TEST_IOT_SDIO )
```
Comment (lundinc2) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: Should the global handle be closed in the test tear down as well ?

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 16]
```diff
@@ -0,0 +1,1073 @@
...
+    xSdioHandle = iot_sdio_open(uctestIotSdioValidHostIdx,
+                                uctestIotSdioInvalidSlotIdx);
+    TEST_ASSERT_EQUAL( NULL, xSdioHandle );
+
+    /* Open with invalid host index and invalid card slot index returns NULL. */
+    xSdioHandle = iot_sdio_open(uctestIotSdioInvalidHostIdx,
+                                uctestIotSdioInvalidSlotIdx);
+    TEST_ASSERT_EQUAL( NULL, xSdioHandle );
+
+    /* Try to close a null handle should retrun INVALID_VALUE. */
```
Comment (lundinc2) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: nit: typo on return

File: libraries/abstractions/common_io/test/test_iot_sdio.c [hunk 17]
```diff
@@ -0,0 +1,1073 @@
...
+ */
+TEST( TEST_IOT_SDIO, AFQP_IotSdioDoubleOpenDoubleClose )
+{
+    IotSdioSlotHandle_t xSdioHandle = NULL;
+    IotSdioSlotHandle_t xSdioHandle2 = NULL;
+
+    int32_t lRetVal;
+
+    /* Open with valid host index and slot index should return a non NULL handle. */
+    xSdioHandle = iot_sdio_open(uctestIotSdioValidHostIdx,
```
Comment (lundinc2) [inline libraries/abstractions/common_io/test/test_iot_sdio.c]: Can you run uncrustify on these two files, it does not conform to style guide

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 15 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: The global is closed (is open) in test setup so the tests can open SDIO to run test.  When the tests are done, there is no need to check it again.

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 16 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: done

File: libraries/abstractions/common_io/test/test_iot_sdio.c (hunk 17 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/test/test_iot_sdio.c]: done

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 10 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: The 2 boolean inputs are separate parameters:
typedef struct IoTSdioPowerOnOffCard
{
    bool bPowerOn;     /*!< True to power on card, false to power off card. */
    bool bUseCallback; /*!< indicate if user callback should be used to power on/off card. */
} IoTSdioPowerOnOffCard_t;

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 9 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: done

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 11 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: done

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 12 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: done

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 14 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: done

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 13 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: done

**Comment by qizengqizeng** [comment]
> I am happy with the code here, will approve this PR once the spelling and phrasing has all beed addressed.

Done.

**Pull request merged**